File: ios-native/ClubFuoco/Features/InviteClaim/InviteHandoff.swift
import Foundation
import UIKit

/// Recovers an invite that was tapped BEFORE the app existed on this phone.
///
/// A Universal Link only opens an installed app. Someone who taps /i/<token>
/// without Club Fuoco goes to the App Store and cold-launches here knowing
/// exactly what they wanted and holding nothing that says so. Apple provides no
/// deferred deep link — App Store campaign parameters are not readable by the
/// app — so the web page pushes the invite through two channels and this reads
/// whichever arrived:
///
///   1. CLIPBOARD — deterministic. `detectPatterns` answers "is there a URL?"
///      WITHOUT prompting; only if it says yes do we read, which is the one
///      moment the system paste prompt makes sense to the person seeing it.
///   2. SERVER TICKET — probabilistic. /api/invite-handoff/claim matches a
///      coarse fingerprint (IP + iOS version) recorded when they left for the
///      App Store.
///
/// THE RULE: this only ever PRE-FILLS the invite sheet. It never claims a spot.
/// Channel 2 can be wrong — two people behind one carrier NAT on the same iOS
/// version inside the 30-minute window will cross — and the containment is that
/// being wrong costs someone one event page they didn't ask for.
///
/// Runs once per install. A guest whose handoff missed still has the link in
/// their messages, and tapping it now works, so there is nothing to retry.
@MainActor
enum InviteHandoff {

    private static let doneKey = "cf.inviteHandoff.checked"

    static func resolveIfNeeded(api: APIClient) async {
        guard !UserDefaults.standard.bool(forKey: doneKey) else { return }
        UserDefaults.standard.set(true, forKey: doneKey)

        // Already handled — a Universal Link fired before we got here, which
        // means the app WAS installed and none of this applies.
        guard InviteLinkRouter.shared.pendingToken == nil else { return }

        if let token = tokenFromClipboard() {
            InviteLinkRouter.shared.pendingToken = token
            return
        }
        if let token = await tokenFromServer(api: api) {
            InviteLinkRouter.shared.pendingToken = token
        }
    }

    // MARK: - Channel 1: the clipboard

    /// Reads the pasteboard ONLY when it holds something shaped like a URL.
    ///
    /// `detectPatterns` is the prompt-free question. Skipping it and reading
    /// directly would fire the system paste alert on every first launch,
    /// including for the majority who arrived with no invite at all.
    private static func tokenFromClipboard() -> String? {
        let pb = UIPasteboard.general
        guard pb.hasStrings else { return nil }
        guard let url = pb.url ?? URL(string: pb.string ?? "") else { return nil }
        return inviteToken(from: url)
    }

    /// `https://clubfuoco.com/i/<token>` → `<token>`.
    ///
    /// Host-checked: the clipboard is whatever the user last copied, so a
    /// stranger's link — or anything else with a /i/ path — must not be treated
    /// as ours.
    static func inviteToken(from url: URL) -> String? {
        guard let host = url.host?.lowercased(),
              host == "clubfuoco.com" || host == "www.clubfuoco.com" else { return nil }
        let parts = url.path.split(separator: "/", omittingEmptySubsequences: true)
        guard parts.count >= 2, parts[0] == "i" else { return nil }
        let token = String(parts[1])
        return token.isEmpty ? nil : token
    }

    // MARK: - Channel 2: the server ticket

    private static func tokenFromServer(api: APIClient) async -> String? {
        // APIClient encodes with .convertToSnakeCase, so `osVersion` goes out
        // as `os_version` — which is what the route reads.
        struct Body: Encodable { let osVersion: String }
        struct Resp: Decodable { let token: String? }
        do {
            // The full version string; the SERVER trims it to major.minor,
            // because Safari reports "18_5_1" on a point release while
            // UIDevice says "18.5.1". Both sides normalise identically there,
            // and doing it in one place keeps them from drifting apart.
            let resp: Resp = try await api.post(
                "/api/invite-handoff/claim",
                body: Body(osVersion: UIDevice.current.systemVersion))
            guard let token = resp.token, !token.isEmpty else { return nil }
            return token
        } catch {
            // No network on first launch, or nothing matched. Either way the
            // link is still in their messages.
            return nil
        }
    }
}

File: src/lib/invite-handoff.ts
import { createHash } from 'crypto'

// ── Deferred invite links ────────────────────────────────────────────────────
//
// A Universal Link only opens the app if the app is already installed. Someone
// tapping /i/<token> without it goes: branded page → App Store → install →
// cold launch with NO token. The intent dies at the last step, and they land in
// a generic signup wizard with no idea what they were doing.
//
// Apple provides nothing for this. App Store `pt`/`ct` campaign parameters are
// not readable by the app, and IDFA-style fingerprinting is gone. Every
// solution is either EXPLICIT (the user hands us the token — the clipboard
// channel, handled on the client) or PROBABILISTIC (we guess). This file is the
// guess.
//
// THE RULE THAT MAKES A GUESS ACCEPTABLE: a handoff only ever PRE-FILLS. It
// never claims a spot, never joins a list, never spends money. The worst
// outcome of a wrong match is that someone sees an event page they didn't ask
// for and closes it. Nothing here should ever be wired to an action that has
// consequences.

/** How long a ticket stays claimable. Long enough to download an app on a bad
 *  connection, short enough that the collision window stays small. */
export const HANDOFF_TTL_MINUTES = 30

/**
 * The two sides of the match see different worlds, which constrains this
 * heavily:
 *
 *  - the WEB side is Mobile Safari, and its UA string is nothing like the app's
 *  - the APP side has no idea what the browser's UA was
 *
 * So the fingerprint can only use what BOTH can state: the server-observed IP,
 * the platform, and the iOS version — which Safari puts in its UA as
 * "CPU iPhone OS 18_5 like Mac OS X" and which the app reads from
 * UIDevice.current.systemVersion.
 *
 * That is coarse on purpose. It is not an identity and must never be treated as
 * one; two people behind the same carrier NAT on the same iOS version within
 * the TTL will collide. See the rule above for why that is survivable.
 */
export function fingerprint(ip: string, osVersion: string | null): string {
  // Peppered so the stored value isn't a reversible list of visitor IPs. Not a
  // security boundary — the table is service-role only — just hygiene.
  const pepper = process.env.INVITE_HANDOFF_SECRET ?? 'fuoco-invite-handoff-v1'
  return createHash('sha256')
    .update([pepper, ip.trim(), 'ios', osVersion ?? 'unknown'].join('|'))
    .digest('hex')
}

/**
 * "18.5" out of a Mobile Safari user-agent, or null.
 *
 * Only major.minor: iOS reports "18.5" from systemVersion while Safari's UA can
 * carry a patch segment ("18_5_1"), and a mismatch there would fail every match
 * for anyone on a point release.
 */
export function osVersionFromUA(ua: string | null): string | null {
  if (!ua) return null
  const m = ua.match(/(?:iPhone|CPU) OS (\d+)[._](\d+)/i)
  if (!m) return null
  return `${m[1]}.${m[2]}`
}

/** Same normalisation for the version the app reports, so both sides agree. */
export function normaliseOsVersion(raw: unknown): string | null {
  if (typeof raw !== 'string') return null
  const m = raw.trim().match(/^(\d+)[._](\d+)/)
  return m ? `${m[1]}.${m[2]}` : null
}

/**
 * Is this request plausibly an iPhone?
 *
 * A ticket is only useful to an iOS install, and refusing to record anything
 * else keeps desktop and crawler traffic out of the match pool — every row that
 * can't match is a row that can only cause a false one.
 */
export function looksLikeIOS(ua: string | null): boolean {
  return !!ua && /iPhone|iPad|iPod/i.test(ua)
}
